src/main/cpp/_nix_based/jssc.cpp

```diff
@@ -529,14 +529,57 @@ JNIEXPORT jboolean JNICALL Java_jssc_SerialNativeInterface_writeBytes
   (JNIEnv *env, jobject, jlong portHandle, jbyteArray buffer){
     jbyte* jBuffer = env->GetByteArrayElements(buffer, JNI_FALSE);
     jint bufferSize = env->GetArrayLength(buffer);
-    jint result = write(portHandle, jBuffer, (size_t)bufferSize);
+    fd_set write_fd_set;
+    int byteRemains = bufferSize;
+    struct timeval timeout;
+    int result;
+    jclass threadClass = env->FindClass("java/lang/Thread");
+    jmethodID areWeInterruptedMethod = env->GetStaticMethodID(threadClass, "interrupted", "()Z");
+      
+    while(byteRemains > 0) {
+        
+        // Check if the java thread has been interrupted, and if so, throw the exception
+        if (env->CallStaticBooleanMethod(threadClass, areWeInterruptedMethod)) {
+            jclass excClass = env->FindClass("java/lang/InterruptedException");
+            env->ThrowNew(excClass, "Interrupted while writing to serial port");
+            // It shouldn't matter what we return, the exception will be thrown right away
+            break;
+        }
+        
+        timeout.tv_sec = 0;
+        timeout.tv_usec = 100000; // 100 ms
+        FD_ZERO(&write_fd_set);
+        FD_SET(portHandle, &write_fd_set);
+        
+        result = select(portHandle + 1, NULL, &write_fd_set, NULL, &timeout);
+        if (result < 0) {
+            env->ThrowNew(env->FindClass("java/io/IOException"), "Waiting for serial port to become writable failed");
+            // Return value is ignored anyway, exception is handled immidiatly
+            break;
+        } else if (result == 0) {
+            // timeout
+            continue;
+        }
+        
+        result = write(portHandle, jBuffer + (bufferSize - byteRemains), byteRemains);
+        if(result < 0){
+            env->ThrowNew(env->FindClass("java/io/IOException"), "Error writing to serial port");
+            break;
+        } else if (result == 0) {
+            env->ThrowNew(env->FindClass("java/io/IOException"), "Serial port was closed unexpectedly");
+            break;
+        } else { // result > 0
+            byteRemains -= result;
+        }
+    }
     env->ReleaseByteArrayElements(buffer, jBuffer, 0);
-    return result == bufferSize ? JNI_TRUE : JNI_FALSE;
+    return JNI_TRUE; //result == bufferSize ? JNI_TRUE : JNI_FALSE;
 }
 
 /**
  * Waits until 'read()' has something to tell for the specified filedescriptor.
  */
+/*
```

Comment:
You are completely right, to be honest, we have already commented to the authors of issue #90 that it would be awesome if the `poll()` technique could be used for this proposal.

We never had any problems with `select()` because of to many (open?) file-descriptors. So we never really dived into this problem/solution. But we encourage you to adjust/refactor our code and migrate to using the `poll()` technique.

Reply:
I had an idea how we could reduce the impact without the need for "*poll*". I did not find a good way to show that patch here. So I opened [PR_128] that we can "*see*" the idea. Basically it just replaces the SEGFAULT by a java exception and so at least we get a proper error message. Additionally I removed the now unused stuff around "*poll*".

[PR_128]: https://github.com/java-native/jssc/pull/128